### DISCUSSION TITLE
[Suggestion] Provide more accurate lists of parameters or improve documentation of some specific MagickWand methods

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5966] [Development]
### ImageMagick version

7

### Operating system, version and so on

windows/android/linux/mac/ios

### Description

In an earlier ticket here it was mentioned that MagickWand API is intended to hide the complexities of MagickCore API. But there are some methods in the MagickWand API  that accept or return structs from the MagickCore API like these one:

```
MagickBooleanType MagickConvolveImage(MagickWand *wand, const KernelInfo *kernel);

MagickBooleanType MagickColorMatrixImage(MagickWand *wand, const KernelInfo *color_matrix);

MagickBooleanType MagickMorphologyImage(MagickWand *wand, const MorphologyMethod method,const ssize_t iterations, const KernelInfo *kernel);
``` 
The problem with this method for a user like me who is new to the API, is that the documentation is not clear about which fields of the kernel should be set and which can be left blank or null. For example, should I set the offsets x and y to 0 or some other default value if I don't want to use them? Or should I keep their garbage values unchanged?

Right now my approach to know which fields are required and which are not is trial and error. It is not an easy task to read a complex method that consists of hundreds of lines each time to know which fields of the struct `kernelInfo` may cause some memory exception that will crash my whole app.

Moreover, the documentation of  `MagickMorphologyImage` says that it accepts an array of doubles as a kernel, but actually it accepts a `KernelInfo` struct with alot of fields that the documentation does not specify which of them can be blank and which should not.

So What I suggest for such methods is :
- Either improve the documentation to mention the restrictions on each field and what it does represent exactly
- Change the parameters that come from MagickCore and split them into separate parameters. For example, instead of accepting a whole `KernelInfo` struct we can accept only the fields of `KernelInfo` that should be set by the user, and then, using these values and using the other default values of `KernelInfo` that were not provided by the user, we can construct a `KernelInfo` inside the corresponding method.

Let me know if you are interested in any of the above suggestions.

Since I am creating bindings to the whole MagickWand API (all the functions [here](https://imagemagick.org/script/magick-wand.php)), so I will definitely pass on each and every method in the API, so this is the best chance to track (mention) all such functions in a comment under this ticker so they will be know to be updated or documented in more details.

And just as a note, when I say it is ambiguous what fields of a struct can be null or be left blank and what fields cannot, then of course I have researched what the method does regarding the image processing, but still, these details are hidden in the implementation of the methods and unless they are documented it is a pain to go with the trial/error approach to figure out what is the constraint on each param.

Thank you.

**Comment by dlemstra** [comment]
We cannot change api arguments. But feel free to open pull requests to improve the documentation. And when you have doubts about what the method requires then search through our code for other places where we call that method.
